### PR TITLE
[vdk-core] Add ingestion functional tests

### PR DESCRIPTION
**Pull request by doks5**
As part of https://github.com/vmware/versatile-data-kit/pull/690, the
IngesterRouter and IngesterBase classes were extended to support ingestion
plugin chaining.

Although unit tests were added as part of the aforementioned PR, no functional
tests were implemented.

This change adds functional tests, that verify the new end-to-end ingestion flow
with pre- and post- processors.

Testing Done: functional tests introduced.

Signed-off-by: Andon Andonov <andonova@vmware.com>

### CODE REVIEW

File: projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py [hunk 1]
```diff
@@ -9,68 +9,63 @@
...
-#    job code is D
-# 4) INGEST_PAYLOAD_PREPROCESS_SEQUENCE=""
-# 5) INGEST_PAYLOAD_PREPROCESS_SEQUENCE=A,B - A,B are used for pre-processing,
-#    B is used for ingesting and post-processing
-
 
 log = logging.getLogger(__name__)
 
 
+# INGEST_PAYLOAD_PREPROCESS_SEQUENCE=A,B and VDK_INGEST_METHOD_DEFAULT=C
```
Comment (ivakoleva) [inline projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py:24]: I think test methods need to be sorted by complexity, starting with ingestion scenarios only (no processing), then pre-processing, then post-processing, then entire end-to-end verification

File: projects/vdk-core/tests/functional/ingestion/jobs/ingest-method-job/1_ingest_step.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+from vdk.api.job_input import IJobInput
+
+
+def run(job_input: IJobInput):
```
[the remark this answers is not in the document]
Comment (doks5) [reply projects/vdk-core/tests/functional/ingestion/jobs/ingest-method-job/1_ingest_step.py]: Indeed, there is no point in the ingest-method-job. Removed, thanks

File: projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py (hunk 1 above)
Comment (doks5) [reply projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py]: Tests sorted. Thanks

File: projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py [hunk 3]
```diff
@@ -9,95 +9,227 @@
...
+    assert ingest_plugin.payloads[1].payload[0] == expected_rows_object
+    assert ingest_plugin.payloads[1].destination_table == "tabular_table"
+
+
+# INGEST_PAYLOAD_PREPROCESS_SEQUENCE=A,B no ingest method specified
+@mock.patch.dict(
+    os.environ,
+    {"VDK_INGEST_PAYLOAD_PREPROCESS_SEQUENCE": "convert-to-string," "add-payload-size"},
+)
+def test_chained_ingest_no_method_passed():
```
Comment (tozka) [inline projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py:96]: In that case would the preprocess plugins be invoked ? 

I think it's fine either way as long as it's documented and we verification in the test here whichever it is .

**Review by tozka — APPROVED** [review]
Looks good to me. I made a very small comment about one test. But ship it after you address it.

File: projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py [hunk 4]
```diff
@@ -9,95 +9,227 @@
...
+
+# INGEST_PAYLOAD_PREPROCESS_SEQUENCE=A,B - A,B are used for pre-processing,
+#    B is used for ingesting and post-processing
+@mock.patch.dict(
+    os.environ,
+    {
+        "VDK_INGEST_PAYLOAD_PREPROCESS_SEQUENCE": "convert-to-string,"
+        "add-payload-size",
+        "VDK_INGEST_METHOD_DEFAULT": "add-payload-size",
+        "VDK_INGEST_PAYLOAD_POSTPROCESS_SEQUENCE": "add-payload-size",
```
Comment (ivakoleva) [inline projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py:206]: would removing the post-processing sequence change the test result?

File: projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py [hunk 5]
```diff
@@ -9,95 +9,227 @@
...
+
+    # Use a sample data job, in which `method` argument is not passed to
+    # ingestion method calls.
+    result: Result = runner.invoke(
+        ["run", jobs_path_from_caller_directory("no-ingest-method-job")]
+    )
+
+    cli_assert_equal(0, result)
+
+    expected_object = dict(
```
Comment (ivakoleva) [inline projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py:222]: It seems a lot of duplicated 
```
expected_object = dict(
        int_key=str(42),
        str_key="example_str",
        bool_key=str(True),
        float_key=str(1.23),
        nested=str(dict(key="value")),
        payload_size=163,
    )
```
and 
```
expected_rows_object = {"first": "two", "payload_size": 80, "second": "2"}
```
used by different tests; maybe define in one place then reuse?

File: projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py (hunk 3 above)
Comment (doks5) [reply projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py:102]: No, pre-processors will not be invoked. I've added a note about this behaviour. Thanks

File: projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py (hunk 4 above)
Comment (doks5) [reply projects/vdk-core/tests/functional/ingestion/test_ingest_chain.py:198]: Not at this point in time. I've added a TODO to modify the behaviour of the test plugin used for post-processing to expose the result of post-processing, in order for us to verify operations. For the time being, the tests in test_ingester_base handle such checks.

**Review by ivakoleva — APPROVED** [review]

**Pull request merged**